Introduce map clone method

adonath opened the pull request:
This PR introduces the `Map.clone()` method to simplify creating new maps from existing ones. This PR addresses #1547.

registerrier commented on gammapy/maps/hpxnd.py:
```diff
@@ -204,17 +204,10 @@ def sum_over_axes(self):
+        geom = self.geom.to_image()
+        axis = tuple(range(self.data.ndim - 1))
+        data = np.nansum(self.data, axis=axis)
+        return self.clone(geom=geom, data=data)
```
Why did you remove the case for multi resolution map? Does this implementation work for healpix maps with mutliresolution scales?
Again the logic of cloning with a new geom is a bit strange to me. You just keep `unit` and `meta`. You don't gain much in terms of complexity with: `map_out = self.__class__(geom=geom, meta=meta, unit=unit)`

registerrier commented on gammapy/maps/hpxnd.py:
```diff
@@ -278,31 +271,28 @@ def pad(self, pad_width, mode='constant', cval=0.0, order=1):
     def downsample(self, factor, preserve_counts=True):
+        geom = self.geom.downsample(factor)
+        coords = self.geom.get_coord()
+        vals = self.get_by_coord(coords)
```
Isn't it slower to use `get_by_coord` rather than `get_by_idx` used in the initial implementation?

registerrier commented on gammapy/maps/base.py:
```diff
@@ -895,6 +897,38 @@ def set_by_idx(self, idx, vals):
+
+        if 'geom' in kwargs and 'data' not in kwargs:
+            raise ValueError("Can't clone and overwrite geometry if the data is not overwritten too.")
+
```
OK I see this is actually most of the usage for this `clone`method. 
I think then this should be a private method. In my opinion, it is very confusing for a user to have a clone that basically can recreate a `Map` from scratch e.g. `map.clone(geom=geom, data=data, meta=meta, unit=unit)`

adonath commented:
I actually noticed the same issue, that the `.clone()` method should not change the geometry of the map. Unfortunately most of the internal usages do exactly that. Which is a bit of a pity, because we loose most of the elegance and simplification we wanted to introduce with the `.clone()` method. So I guess for those cases I'll change back to the old pattern with `self.__class__()` and just make sure, that I do it once correctly and uniformly. Do you agree @cdeil and @registerrier?

adonath commented:
I've changed now to the following pattern. I introduced a hidden `_init_copy()` method, that is used to get a clean implementation of all the cases where again a `Map` object of the same type returned. In addition there is a public `.copy()` method, which forbids to change the map geometry. I must say I', still not happy with the pattern, but I don't want to spend any time further on this.

cdeil commented:
I don't like the `_init_copy` in addition to `copy`, because then half of the time we'll use one and half of the time the other, i.e. devs have to learn both, for no good reason.
Also the `_init_copy` still does the `__init__` argument inspecition, which I would like to avoid, again just for the reason "keep it simple", Gammapy code should be dead simple everywhere.

I'll make a follow-up commit here and then merge without further discussion, so that we can move on.

That doesn't have to be the final solution, we can re-discuss on Friday or later once we have more experience and callers with this, and then put something better if we find it.

But this should go in, IMO it's been a real pain-point and error-prone, to not have a way to make a copy without having to import and hard-code the map class (i.e. WCSNDMap or HPX map).
